src/app_engine/gcm_notify.py

```diff
@@ -44,94 +45,119 @@ def get_gcm_api_key():
+                    status_code, payload, result.content)
+      return
+    # It's possible to get a 200 but receive an error from GCM server. This will
+    # be recorded in the response.
+    logging.info('GCM request result:\n%s', result.content)
+
+  # Create async requests.
   headers = {
       'Content-Type': 'application/json',
       'Authorization': ('key=%s' % gcm_api_key)
   }
-  payload = create_gcm_payload(gcm_ids, collapse_key, message)
+  rpcs = []
   start_time = time.time()
-  # TODO(tkchin): investigate setting follow_redirects=False.
-  result = urlfetch.fetch(url=GCM_API_URL,
-                          payload=payload,
-                          method=urlfetch.POST,
-                          headers=headers)
-  end_time = time.time()
-  logging.info('Took %.3fs to send GCM message request with %d endpoints. ' +
-               'Payload = %s',
-               end_time - start_time,
-               len(gcm_ids),
-               payload)
-  status_code = result.status_code
-  if status_code != 200:
-    logging.error('Failed to send GCM message. Result:%d', status_code)
-    logging.error('Response: %s', result.content)
-    return
-  # It's possible to get a 200 but receive an error from GCM server. This will
-  # be recorded in the response.
-  logging.info('GCM request result:\n%s', result.content)
-
-
-def create_gcm_payload(gcm_ids, collapse_key, data):
-  return json.dumps({
-      'data': data,
-      'registration_ids': gcm_ids,
-      'collapse_key': collapse_key,
-      'time_to_live': GCM_TIME_TO_LIVE_IN_SECONDS
-  })
-
-
-def create_invite_message(room_id, caller_id, metadata):
-  return {
-      GCM_MESSAGE_TYPE_KEY: GCM_MESSAGE_TYPE_INVITE,
-      constants.PARAM_ROOM_ID: room_id,
-      constants.PARAM_CALLER_ID: caller_id,
-      constants.PARAM_METADATA: metadata,
-  }
-
-
-def send_invites(gcm_ids, room_id, caller_id, metadata):
-  """Create and send the invite message to the callee.
-
-  Args:
-    gcm_ids: ids to send the message to.
-    room_id: id of the room to join.
-    caller_id: the callable public id of the caller.
-    metadata: value passed through from caller to callee. Used for call
-        specific data that the server does not need to know about, such
-        as creating an audio only call.
-
-  Returns:
-    The result of send_gcm_messages.
-
-  """
-  message = create_invite_message(room_id, caller_id, metadata)
-  return send_gcm_messages(gcm_ids, message, room_id)
-
-
-def create_bye_message(room_id, reason):
-  return {
-      GCM_MESSAGE_TYPE_KEY: GCM_MESSAGE_TYPE_BYE,
-      constants.PARAM_ROOM_ID: room_id,
-      GCM_MESSAGE_REASON_KEY: reason
-  }
-
-
-def send_byes(gcm_ids, room_id, reason):
-  message = create_bye_message(room_id, reason)
-  return send_gcm_messages(gcm_ids, message, room_id)
+  for message in messages:
+    payload = message.get_gcm_payload()
+    rpc = urlfetch.create_rpc()
+    rpc.callback = lambda: handle_rpc_result(rpc, payload)
```

Comment:
Haha, I'm biased because my first real language was LISP. Leaving it since you're okay with it.